lib/names/workloadendpoint.go

```diff
@@ -0,0 +1,217 @@
+func (ids WorkloadEndpointIdentifiers) NameMatches(name string) (bool, error) {
+	// Extract the required segments for this orchestrator type.
+	req, err := ids.getSegments()
+	if err != nil {
+		return false, err
+	}
+
+	// Extract the parameters from the name.
+	parts := extractDashSeparatedParms(name, len(req))
+	if len(parts) == 0 {
+		return false, nil
+	}
+
+	// Check each name segment for a non-match.
+	for i, r := range req {
+		if r.value != "" && r.value != parts[i] {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+// CalculateWorkloadEndpointName calculates the expected name for a workload
+// endpoint given the supplied Spec.  Calico requires a precise naming convention
+// for workload endpoints that is based on orchestrator and various other orchestrator
+// specific parameters.
+//
+// If allowPrefix is true, we construct the name prefix up to the last specified index
+// and terminate with a dash.
+func (ids WorkloadEndpointIdentifiers) CalculateWorkloadEndpointName(allowPrefix bool) (string, error) {
+	req, err := ids.getSegments()
+	if err != nil {
+		return "", err
+	}
+
+	parts := []string{}
+	for _, s := range req {
+		part := ""
+
+		if len(s.value) == 0 {
+			// This segment has no value associated with it.
+			if !allowPrefix {
+				// We are not allowing prefixes.  This is an error scenario
+				return "", cerrors.ErrorValidation{
+					ErroredFields: []cerrors.ErroredField{
+						{Name: s.field, Value: s.value, Reason: "field should be assigned"},
+					},
+				}
+			}
+
+			// We are allowing prefixes, so return the prefix that we have constructed thus far,
+			// terminating with a "-".
+			return strings.Join(parts, "-") + "-", nil
+		}
+
+		part, ef := escapeDashes(s)
+		if ef != nil {
+			return "", cerrors.ErrorValidation{ErroredFields: []cerrors.ErroredField{*ef}}
+		}
+		parts = append(parts, part)
+	}
+
+	// We have extracted all of the required segments, join the segments with a "-" and
+	// return that as the name.
+	return strings.Join(parts, "-"), nil
+}
+
+// getSegments returns the ID segments specific to the orchestrator.
+func (ids WorkloadEndpointIdentifiers) getSegments() ([]segment, error) {
+	node := segment{value: ids.Node, field: "node", structField: "Node"}
+	orch := segment{value: ids.Orchestrator, field: "orchestrator", structField: "Orchestrator"}
+	cont := segment{value: ids.ContainerID, field: "containerID", structField: "ContainerID"}
+	pod := segment{value: ids.Pod, field: "pod", structField: "Pod"}
+	endp := segment{value: ids.Endpoint, field: "endpoint", structField: "Endpoint"}
+	workl := segment{value: ids.Workload, field: "workload", structField: "Workload"}
+
+	// Node and Orchestrator are *always* required.
+	if len(node.value) == 0 {
+		return nil, cerrors.ErrorValidation{
+			ErroredFields: []cerrors.ErroredField{
+				{Name: node.field, Reason: "field should be assigned"},
+			},
+		}
+	}
+	if len(orch.value) == 0 {
```

Comment:
shouldn't we also be making sure that the `endpoint` is specified since that's also required for all the orchestrators?

Reply:
nope - it's the last part in the name construction, so for prefixes it won't be specified.  For non-prefixes it'll be caught when we construct the name from the segments, because that does a final validation of all parts.